src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp

```diff
@@ -631,10 +633,23 @@ RangesInDataParts MergeTreeDataSelectExecutor::filterPartsByPrimaryKeyAndSkipInd
     bool use_skip_indexes,
     bool find_exact_ranges)
 {
-    RangesInDataParts parts_with_ranges;
-    parts_with_ranges.resize(parts.size());
     const Settings & settings = context->getSettingsRef();
 
+    if (context->canUseParallelReplicasOnFollower() && settings[Setting::parallel_replicas_local_plan]
```

Comment:
it should work, but there is a caveat. some logic (a.k.a. optimisations) depend on `sum_marks` and if replicas will see much bigger number of marks smth might degrade. not sure how significant that is.

Reply:
AFAIS, the `selected_marks` is used only for projection optimisations but they are disabled for queries with parallel replicas - https://github.com/ClickHouse/ClickHouse/blob/cb582abbb51e546a5dc108e76182b6ed57c60b07/src/Processors/QueryPlan/Optimizations/projectionsCommon.cpp#L46-L47

Reply:
we also use it here:

https://github.com/ClickHouse/ClickHouse/blob/5e2975baa5a7d45b2f74fbb6b8a1847f2fd1efdc/src/Storages/MergeTree/MergeTreeReadPoolParallelReplicas.cpp#L14-L15

and here:

https://github.com/ClickHouse/ClickHouse/blob/5e2975baa5a7d45b2f74fbb6b8a1847f2fd1efdc/src/Storages/MergeTree/MergeTreeReadPoolBase.cpp#L62-L67